### ISSUE TITLE
perf(vulkan): one prefill fence instead of per-layer fence round-trips

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#13043]
# perf(vulkan): one prefill fence instead of per-layer fence round-trips

```routing
lane: pkg/engine
paths: ["pkg/engine/vulkan_qwen35_sequence.go"]
expected_steps: 5
cross_key: spob-w2-vulkan-prefill-layer-fence
```

## Core through-line
`pkg/engine/vulkan_qwen35_sequence.go:623` (`Qwen35SequencePrefill`) issues
`C.fvk_batch_flush_status()` inside `for i, layer := range req.Layers` (line 560),
where `fvk_batch_flush_status` → `batchFlush()` → `endSubmitWait` →
`vkQueueSubmit` + `vkWaitForFences(..., UINT64_MAX, VK_TRUE)` (full host block) per
layer. For the dense Qwen3.8 27B that is 64 host↔GPU round-trips per prefill, even
though the whole prefill is already bracketed by `fvk_batch_begin()` (:496) and a
single `final-fence` (:648).

## Three-conjunct witness (SPOB)
- **Serial blocking:** each layer waits on `vkWaitForFences` before the next.
- **Hot-loop multiplier:** 64 layers per prefill.
- **Collapsible:** the per-layer flush only bounds scratch lifetime
  (`qwen35VulkanSequenceReleaseLocked`); deferring scratch reclamation to the
  end-of-prefill boundary (or replaying a `vulkan_graph.go` layer graph) collapses
  64 submit+fence round-trips to 1, matching the existing final fence.

## Current state
Per-layer fence in the dense prefill sequence.

## Design
- Defer scratch release to the prefill boundary and flush once, or use the shipped `vulkan_graph.go` reusable layer graph.

## Gold-plating boundary
- No change to output/logits semantics.
- No new backend; reuse `vulkan_graph.go` / the existing final fence.
- Public `fak` — file in `anthony-chaudhary/fak`.

### File:Line Seam
`pkg/engine/vulkan_qwen35_sequence.go:623`; final fence `:648`; graph sibling `pkg/engine/vulkan_graph.go`.

## Likely files
- `pkg/engine/vulkan_qwen35_sequence.go`

## Verifiable Witness
```bash
go -C .. test ./pkg/engine/... -run Vulkan -count=1
```

## Done condition / witness
Witness: a fence-counting test shows one prefill flush (plus final) instead of
per-layer, with identical logits.

### Blast Radius & Affected Lanes
- **Primary Lane:** `pkg/engine` (public fak)
- **Unaffected:** CUDA/Metal paths; boundary check exit 0

### Quarantined Fallback Mechanism
Keep per-layer fencing behind a flag until the batched path is witnessed on hardware.

## Scoped Acceptance Criteria
- [ ] Per-layer fence replaced by a single boundary flush (or graph replay).
- [ ] Identical logits (parity test).
- [ ] Scratch lifetime safety preserved.
- [ ] `gofmt`/`go vet` clean.

## Working spine
Begin batch → encode all layers → release scratch at boundary → single fence → final.

## Acceptance gate
`go -C ../fak test ./pkg/engine/... -count=1` exits zero.

## Closure binding
The resolving commit cites this issue (`Fixes #<n>`) and carries `(fak compute)`.

## Parent context
Parent: #1372 (shift-left perf-patterns)
Theme: [`shift-left-perf-patterns/INDEX.md`](../../tickets/shift-left-perf-patterns/INDEX.md) SP15 (other-backend lanes).
Class: `docs/notes/CONCEPT-SERIAL-ROUND-TRIP-PATTERN-CLASS-2026-09-13.md` §3 (CUDA/ROCm row).
Wave 2 audit: `docs/goals/GOAL-spob-audit-wave2-20260914.md` (public fak lane).
## Value
- Centrality: Core
- P1 Context: advanced - collapses serial round-trips so per-operation context is reused instead of re-fetched
- P2 Net value: advanced - removes N-1 blocking round-trips from a hot path where a batch/fuse sibling exists
- P3 Adaptation: N/A - this is a fixed mechanism fix, not an adaptive-policy change
- P4 Operations: preserved - existing operator path and output contract remain intact

## Work estimate
Estimate: 5 points (small-medium; one seam, one collapsible sibling already named)

## Overall completion contribution
Contribution: 5/40 points

## Completion standard
production

## Target operating envelope
- roundtrip_per_hot_op: >= 1 batch

## Witnessed operating envelope
- roundtrip_per_hot_op: >= 1 batch

## Work unit
One SPOB seam: replace serial per-op blocking round-trips with the named collapsible batch/fuse sibling.

## Assumptions
- The named collapsible sibling is reachable from the hot path and preserves output semantics.
- Per-op vs batched equivalence is testable with a counting/fake executor.

## Confusion risks
- The collapsible sibling may not cover every edge case (error handling, partial failure); keep a per-op fallback.
- "Serial" may be intentional for safety in some loops — this issue names only the confirmed hot-path cases.

## Coordination
- Owns only the named files/lane; disjoint from peer SPOB tickets (#1450-#1454, SP6-SP16).
- No overlap with worktree census ticket #1442 except the explicitly noted adjacent paths.

## Trigger
SPOB class audit wave 2 (2026-09-14) — missed instance found by deep cross-domain scan.

## Batch policy
One issue per confirmed SPOB seam; deduped against existing SPOB tickets by key.

## Why this is next
SPOB class audit wave 2 (2026-09-14) found this instance by deep cross-domain scan;
it was missed by the 2026-09-13 survey (SP6-SP16) despite a collapsible sibling already
existing in-tree, so it is shift-left prevention work on the BB-perf-01 keystone class.

**Comment by anthony-chaudhary** [comment]
﻿Resolved on trunk in commit b3e52750b (`perf(compute): collapse the per-layer prefill fence into the boundary flush`, `(fak compute)`).

**What landed:** `Qwen35SequencePrefill` (`internal/compute/vulkan_qwen35_sequence.go`) no longer issues one `fvk_batch_flush_status()` submit+blocking-fence per layer (64 per dense prefill). The default policy records every layer into the loop-entry `fvk_batch_begin()` batch and flushes once at the existing boundary fence; `FAK_QWEN35_SEQUENCE_PER_LAYER_FENCE=1` restores the legacy per-layer round-trip for a matched A/B. New tag-free policy file `internal/compute/qwen35_sequence_fence_policy.go` + test.

**Safety:** the release path calls raw `C.fvk_free`, which parks scratch in `g_batchFreed` while `g_batching` is true, so nothing is handed back out until the boundary `endSubmitWait` fences the single submit — no WAR/UAF, no leak (adversarially cross-checked against the C++ shim). Residual is a bounded footprint increase (one layer-set of parked buffers held to the boundary).

`[SW-VERIFIED]` witnesses (2026-09-17, trunk b3e52750b):
- `go test ./internal/compute/ -run TestQwen35VulkanSequencePerLayerFencePolicy -count=1` → PASS
- `go test -race ./internal/compute/ -run TestQwen35VulkanSequencePerLayerFencePolicy -count=1` → ok 1.269s
- `go test ./internal/compute/ -count=1` → ok 84.797s
- `go build ./internal/...` → exit 0; `go build -tags vulkan ./internal/compute/` → exit 0
- `go vet -tags vulkan ./internal/compute/` → clean; `gofmt -l` → empty
- `go run ./cmd/fak-boundary check --staged` → 5/5 gates PASS, 0 violations

**Residual (NOT claimed):** the `[HW-WITNESSED]` "identical logits on device" criterion still needs a Vulkan-tagged parity run on gfx1151. That run is environment-blocked on this host: the `-tags vulkan` **test** binary fails to link (`libfakvulkan.a` + mingw .drectve mismatch), and it fails **identically on unmodified trunk**, so it is pre-existing toolchain state, not a regression from this change. The collapsed default's on-device logits parity and the round-trip payoff remain to be witnessed.

**Comment by anthony-chaudhary** [comment]
Landed on trunk in b3e52750b (fak compute). Fence collapse is [SW-VERIFIED]; on-device identical-logits parity remains [HW-WITNESSED] and is flagged in the witness comment above.
